### PR TITLE
Update shared vpc config for Project Factory and Project module

### PR DESCRIPTION
**Pull request by simonebruzzechesse**
Update shared vpc config for project factory and project module for more granular Shared VPC configuration. Added the following attributes to shared_vpc_service_config variable:

- subnets_iam: for subnet level IAM binding for non API services identities
- host_project_iam: for granting network user role to identities other than API services

I applicable, I acknowledge that I have:
- [X] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [X] Ran `terraform fmt` on all modified files
- [X] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [X] Made sure all relevant tests pass

### CODE REVIEW

**Comment by ludoo** [comment]
wow you're fast :)

File: modules/project/variables.tf [hunk 1]
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
   description = "Configures this project as a Shared VPC service project (mutually exclusive with shared_vpc_host_config)."
   # the list of valid service identities is in service-agents.yaml
   type = object({
-    host_project                = string
-    service_identity_iam        = optional(map(list(string)), {})
-    service_identity_subnet_iam = optional(map(list(string)), {})
-    service_iam_grants          = optional(list(string), [])
+    host_project                 = string
+    host_project_iam             = optional(list(string), [])
```
Comment (ludoo) [inline modules/project/variables.tf:306]: I would restrict to only assign compute.networkUser, and rename to host_network_users

File: modules/project/variables.tf [hunk 2]
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
...
+    service_identity_iam         = optional(map(list(string)), {})
+    service_identity_subnets_iam = optional(map(list(string)), {})
+    service_iam_grants           = optional(list(string), [])
+    subnets_iam                  = optional(map(list(string)), {})
```
Comment (ludoo) [inline modules/project/variables.tf:310]: same as above, only network user role makes sense from service projects, and for service identities we already support more roles

**Comment by simonebruzzechesse** [comment]
> wow you're fast :)

Not that fast.. I started workingon this yesterday! :)

File: modules/project/variables.tf (hunk 1 above)
Comment (ludoo) [reply modules/project/variables.tf:306]: let's just rename to

- `network_users`
- `network_subnet_users`

File: modules/project/shared-vpc.tf [hunk 3]
```diff
@@ -55,19 +55,37 @@ locals {
...
       for service in services : [{
         region  = split("/", subnet)[0]
         subnet  = split("/", subnet)[1]
         service = service
       }]
     ]
   ])
-  svpc_service_subnet_iam = {
-    for v in local._svpc_service_subnet_iam :
+  svpc_service_subnets_iam = {
```
Comment (ludoo) [inline modules/project/shared-vpc.tf:67]: `svpc_service_subnet_iam`

File: modules/project/shared-vpc.tf [hunk 4]
```diff
@@ -55,19 +55,37 @@ locals {
     "${b.role}:${b.service}" => b
   }
   # normalize the service identity subnet IAM bindings
-  _svpc_service_subnet_iam = flatten([
-    for subnet, services in local._svpc.service_identity_subnet_iam : [
+  _svpc_service_subnets_iam = flatten([
```
Comment (ludoo) [inline modules/project/shared-vpc.tf:58]: `_svpc_service_subnet_iam`

File: modules/project/variables.tf [hunk 5]
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
...
   # the list of valid service identities is in service-agents.yaml
   type = object({
-    host_project                = string
-    service_identity_iam        = optional(map(list(string)), {})
-    service_identity_subnet_iam = optional(map(list(string)), {})
-    service_iam_grants          = optional(list(string), [])
+    host_project                 = string
+    host_project_iam             = optional(list(string), [])
+    service_identity_iam         = optional(map(list(string)), {})
+    service_identity_subnets_iam = optional(map(list(string)), {})
```
Comment (ludoo) [inline modules/project/variables.tf:308]: `service_identity_subnet_iam`

File: modules/project/variables.tf (hunk 1 above)
Comment (ludoo) [inline modules/project/variables.tf:306]: Using a list for type will make it very fragile as dynamic values in keys usually trigger an error on first apply, which was one of the reasons for not supporting this . :) We should mention it in the README (which you might already do, haven't gotten that far yet).

File: blueprints/factories/project-factory/variables.tf [hunk 6]
```diff
@@ -28,10 +28,12 @@ variable "data_defaults" {
...
     services                   = optional(list(string), [])
     shared_vpc_service_config = optional(object({
-      host_project                = string
-      service_identity_iam        = optional(map(list(string)), {})
-      service_identity_subnet_iam = optional(map(list(string)), {})
-      service_iam_grants          = optional(list(string), [])
+      host_project                 = string
+      host_project_iam             = optional(list(string), [])
+      service_identity_iam         = optional(map(list(string)), {})
+      service_identity_subnets_iam = optional(map(list(string)), {})
```
Comment (ludoo) [inline blueprints/factories/project-factory/variables.tf:34]: `service_identity_subnet_iam`

File: blueprints/factories/project-factory/variables.tf [hunk 7]
```diff
@@ -28,10 +28,12 @@ variable "data_defaults" {
...
+      service_iam_grants           = optional(list(string), [])
+      subnets_iam                  = optional(map(list(string)), {})
```
Comment (ludoo) [inline blueprints/factories/project-factory/variables.tf:36]: `subnet_iam`

File: modules/project/README.md [hunk 8]
```diff
@@ -317,9 +317,53 @@ module "service-project" {
 # tftest modules=2 resources=9 inventory=shared-vpc-auto-grants.yaml e2e
 ```
 
-In specific cases it might make sense to selectively grant the `compute.networkUser` role for service identities at the subnet level, and while that is best done via org policies it's also supported by this module.
```
Comment (ludoo) [inline modules/project/README.md:320]: We should keep this sentence, as you're nuking the explanation for service identities. :)

File: modules/project/README.md [hunk 9]
```diff
@@ -317,9 +317,53 @@ module "service-project" {
...
+The 'networkUser' role for identities other than API services (e.g. users, groups or service accounts) can be managed via the `host_project_iam` attribute, by specifying the list of identities.
 
-In this example, Compute service identity will be granted compute.networkUser in the `gce` subnet defined in `europe-west1` region.
+Note that this configuration grants the 'networkUser' role at project level which results in the identity being able to configure resources on all the VPCs and subnets belonging to the host project without further restrictions. It is possible, or better recommended, to restrict which subnets can be used on the newly created project using an Org Policy. The latter applied at project level can restrict access to a list of subnets from the host project. For more information on the Org Policy configuration check the corresponding [Organization Policy section](#organization-policies). The following example details this configuration.
```
Comment (ludoo) [inline modules/project/README.md:322]: `The most reliable way to restrict which subnets [...] is via the xxxxx organization policy.`

Comment (ludoo) [reply modules/project/README.md:322]: ~`The latter applied at project level can restrict access to a list of subnets from the host project.`~

File: modules/project/README.md [hunk 10]
```diff
@@ -317,9 +317,53 @@ module "service-project" {
...
+    host_project     = module.host-project.project_id
+    host_project_iam = ["group:team-1@example.com"]
+    # reuse the list of services from the module's outputs
+    service_iam_grants = module.service-project.services
+  }
+}
+# tftest modules=2 resources=11 inventory=shared-vpc-host-project-iam.yaml e2e
+```
+
+In specific cases it might make sense to selectively grant the `compute.networkUser` role for all kind of identities at the subnet level, although you can restrict access via org policies this is also supported by this module.
```
Comment (ludoo) [inline modules/project/README.md:364]: I would rephrase as

> In specific cases it might make sense to selectively grant the `compute.networkUser` role at the subnet level, and this is also supported by this module.

Then combine the following sentence in the same paragraph.

File: modules/project/README.md [hunk 11]
```diff
@@ -344,12 +388,66 @@ module "service-project" {
...
+    subnets_iam = {
+      "europe-west1/gce" = ["group:team-1@example.com"]
+    }
+  }
+}
+# tftest modules=2 resources=7 inventory=shared-vpc-subnet-grants.yaml
+```
+
+The last example provides a reference Shared VPC configuration that adheres to the principle of least privilege by:
+- Granting API services and other identities roles at the subnet level
```
Comment (ludoo) [inline modules/project/README.md:403]: I would just show the org policy in the project role example above and scrap this example, combining the two does not make a lot of sense IMHO.

Comment (simonebruzzechesse) [reply modules/project/README.md:403]: I added this last example just for showing a "reference" configuration for Shared VPC with permissions assigned at subnet level + org policy which might be a best practise, WDYT?

File: modules/project/README.md (hunk 8 above)
Comment (simonebruzzechesse) [reply modules/project/README.md:320]: I just moved that sentence below the new example section

File: modules/project/README.md (hunk 10 above)
Comment (simonebruzzechesse) [reply modules/project/README.md]: Done!

File: modules/project/shared-vpc.tf (hunk 4 above)
Comment (simonebruzzechesse) [reply modules/project/shared-vpc.tf]: Updated!

File: modules/project/shared-vpc.tf (hunk 3 above)
Comment (simonebruzzechesse) [reply modules/project/shared-vpc.tf]: Updated!

File: modules/project/variables.tf (hunk 1 above)
Comment (simonebruzzechesse) [reply modules/project/variables.tf]: Updated variables and code!

File: modules/project/variables.tf (hunk 5 above)
Comment (simonebruzzechesse) [reply modules/project/variables.tf]: Updated!

File: modules/project/variables.tf [hunk 12]
```diff
@@ -314,10 +316,12 @@ variable "shared_vpc_service_config" {
...
     condition = var.shared_vpc_service_config.host_project != null || (
       var.shared_vpc_service_config.host_project == null &&
+      length(var.shared_vpc_service_config.host_project_iam) == 0 &&
       length(var.shared_vpc_service_config.service_iam_grants) == 0 &&
-      length(var.shared_vpc_service_config.service_iam_grants) == 0
+      length(var.shared_vpc_service_config.service_identity_iam) == 0 &&
+      length(var.shared_vpc_service_config.subnets_iam) == 0
     )
-    error_message = "You need to provide host_project when providing service_identity_iam or service_iam_grants"
+    error_message = "You need to provide host_project when providing shared vpc host and subnets iam permissions."
```
[the remark this answers is not in the document]
Comment (simonebruzzechesse) [reply modules/project/variables.tf]: Updated!

File: blueprints/factories/project-factory/variables.tf (hunk 7 above)
Comment (simonebruzzechesse) [reply blueprints/factories/project-factory/variables.tf]: Replaced with network_subnet_users

File: blueprints/factories/project-factory/variables.tf (hunk 6 above)
Comment (simonebruzzechesse) [reply blueprints/factories/project-factory/variables.tf]: Updated!

File: modules/project/README.md (hunk 9 above)
Comment (simonebruzzechesse) [reply modules/project/README.md]: Updated!

File: modules/project/README.md (hunk 11 above)
Comment (ludoo) [reply modules/project/README.md:401]: you don't need subnet-level with org policy IMHO :) I find this confusing, let's merge the org policy where we show project-level IAM

File: modules/project/variables.tf (hunk 1 above)
Comment (simonebruzzechesse) [reply modules/project/variables.tf]: Added warning for  that attribute! Thanks

**Comment by simonebruzzechesse** [comment]
> Great work Simo, thanks for this. Dropped a few comments mainly on variable naming and the README wording, once those are addressed this is good to go from my PoV.

Thanks for the feedback and comments, I updated the PR with your improvements! WRT what you suggested I just kept the reference example for the Shared VPC being, IMHO, the best configuration which suits most of the use cases. WDYT?

**Comment by ludoo** [comment]
I don't think it's a reference as I wrote twice :) I would just move the org policy to the project-level IAM example, and scrap the last example.

**Review by ludoo — APPROVED** [review]
Thanks Simo!!!
